Added doc for etcd member garbage collection

aaronfern opened the pull request:
**What this PR does / why we need it**:
Added a small document to capture the need and approach taken for etcd member garbage collection

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
To be merger only after #403 

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```improvement operator
NONE
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+
+With support for multi-node operation, there is an expectation that the cluster size can be dynamic. In such scenarios, a reduction in the cluster size should not result in any superfluous members or entries being left over.
+
+Such superfluour members/entries can have a adverse affect on the quorum size of the cluster and could result in a degradation of service.
```
```suggestion
Such superfluour members/entries can have an adverse affect on the quorum size of the cluster and could result in a degradation of service.
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+As such, care has to be taken to periodically check and remove any superfluous entries.
+
+There is the possibility of superfluous entries in the `members` array of the ETCD CRD as well as in the ETCD cluster for which there is no corresponding pod in the `StatefulSet` anymore.
+Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar
```
```suggestion
Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar.
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+
+There is the possibility of superfluous entries in the `members` array of the ETCD CRD as well as in the ETCD cluster for which there is no corresponding pod in the `StatefulSet` anymore.
+Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar
+The scope of this document is the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check
```
```suggestion
The purpose of this document is to describe the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check.
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+
+
+#### Example
+Below is as example of how we could end up with superfluous members in the ETCD cluster.
```
```suggestion
Below is an example of how we could end up with superfluous members in the ETCD cluster.
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
````diff
@@ -0,0 +1,34 @@
+Below is as example of how we could end up with superfluous members in the ETCD cluster.
+This is a common scenario that could occur frequently and demonstrates the need for a garbage collectoin mechanism that detects and clears such superfluous entries automatically
+```
+Suppose an ETCD cluster is scaled up from `3` to `5` and the newly added members were failing constantly due to insufficient resources and then if the ETCD cluster is scaled back down to `3`. The failing member pods may not have the chance to clean up their `member` entries (from the `members` array in the ETCD CRD as well as from the ETCD cluster) leading to superfluous members in the ETCD clust ...
````
I just wondered what we should do if a etcd-member pod receives a SIGTERM which can either mean:
- The pod is removed/evicted
- A rolling update is happening
- The statefulset is scaled down

What we already consider today IIUC:
- Regular garbage collection of superfluous members.
- When a member pod starts, it checks if there is an "outdated" member entry which it is going to replace, i.e. it removes the entry and adds itself again.

:question: Do we also want to have a cleanup logic if SIGTERM arrives or do we solely want to rely on the mechanisms listed above?

Let's discuss this first and then we can adjust this example.

timuthy replied:
As discussed offline, we're going to go with a cleanup logic when SIGTERM arrives.

Can you please delete "from the `members` array in the ETCD CRD" as this is not the case anymore?

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
````diff
@@ -0,0 +1,34 @@
+```
+
+### Superfluous member 
+A member in the etcd cluster can be considered as superfluous only if the following three conditiond hold true
````
```suggestion
A member in the etcd cluster can be considered as superfluous only if the following three conditions hold true:
```

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+### Superfluous member 
+A member in the etcd cluster can be considered as superfluous only if the following three conditiond hold true
+1. It does not have a corresponding pod
+2. It does not have an enter in the ETCD CRD member list (`status.members`)
```
I think this is not relevant anymore, right?

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+2. It does not have an enter in the ETCD CRD member list (`status.members`)
+3. It does not have a corresponding etcd member lease
+
+It may not be necessary to verify all 3 conditions before calling a member superfluous. For instance, `status.members` is kept in sync with member leases by `etcd-druid`.
+Hence, checking the for the correspondng pod and the member entry in `status.members` is a sufficient check
```
```suggestion
```
This is also not relevant anymore, right?

timuthy commented on doc/concepts/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,34 @@
+
+### Approach
+- Superfluous members garbage collection process will only be triggered by the leading `etcd-backup-restore` sidecar
+- The garbage collection mechanish runs periodically
```
Can you describe when and how to influence this?

timuthy reviewed (changes requested):
Thanks for the document. Only some nits and some updates are required after the latest discussions in #403.

ishan16696 commented:
@aaronfern what's the plan here ?
Do we like to add this doc ? Is  etcd member garbage collection already turned on ?

aaronfern commented:
No, etcd member garbage collection is not turned on by default.
However, I'd like it if we can still have this doc that explains the process as it is part of the code

Apologies for this PR being dormant for some time. I will update this PR asap

aaronfern commented:
I have rebased this PR and updated the doc
Although member garbage collection is not enabled by default and we also do not use it in the g/g context, we still do have code for this functionality and hence I would like this doc to be included 

Please review this whenever possible

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+
+With support for multi-node operation, there is an expectation that the cluster size can be dynamic. In such scenarios, a reduction in the cluster size should not result in any superfluous members or entries being left over.
+
+Such superfluous members/entries can have an adverse effect on the quorum size of the cluster and could result in a degradation of service.
+As such, care has to be taken to periodically check and remove any superfluous entries.
```
```suggestion
Such superfluous members/entries can have an adverse effect on the quorum size of the cluster and could result in a degradation of service. As such, care has to be taken to periodically check and remove any superfluous entries.
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+Such superfluous members/entries can have an adverse effect on the quorum size of the cluster and could result in a degradation of service.
+As such, care has to be taken to periodically check and remove any superfluous entries.
+
+There is the possibility of superfluous entries in the `members` array of the ETCD CRD as well as in the ETCD cluster for which there is no corresponding pod in the `StatefulSet` anymore.
+Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar.
+The purpose of this document is the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check
```
```suggestion
There is the possibility of superfluous entries in the `members` array of the Etcd resource as well as in the etcd cluster for which there is no corresponding pod in the `StatefulSet` anymore. Keeping the `members` array on the Etcd resource status in check is the responsibility of `etcd-druid` whereas keeping the etcd cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar. This document describes how the leading `etcd-backup-restore` sidecar keeps the etcd cluster in check.
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+
+
+#### Example
+Below is an example of how we could end up with superfluous members in the ETCD cluster.
+This is a common scenario that could occur frequently and demonstrates the need for a garbage collection mechanism that detects and clears such superfluous entries automatically
```
```suggestion
Below is an example of how we could end up with superfluous members in the etcd cluster. This is a common scenario that could occur frequently and demonstrates the need for a garbage collection mechanism that detects and clears such superfluous entries automatically.
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
````diff
@@ -0,0 +1,30 @@
+Below is an example of how we could end up with superfluous members in the ETCD cluster.
+This is a common scenario that could occur frequently and demonstrates the need for a garbage collection mechanism that detects and clears such superfluous entries automatically
+```
+Suppose an ETCD cluster is scaled up from `3` to `5` and the newly added members are failing constantly due to insufficient resources and then if the ETCD cluster is scaled back down to `3`. The failing member pods may not have the chance to clean up their `member` entries from the ETCD cluster leading to superfluous members in the ETCD cluster that may have an adverse effect on the quorum of the ...
````
```suggestion
Suppose an etcd cluster is scaled up from `3` to `5` replicas, and the newly added members are failing to come up due to insufficient resources. If the etcd cluster is then scaled back down to `3` replicas, the failing member pods may not have the chance to clean up their own `member` entries from the etcd cluster, resulting in superfluous members in the etcd cluster. This directly affects the quorum of the cluster, since failure of either of the 3 existing members will lead to a quorum loss, since the cluster size is still `5` according to the etcd cluster state.
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
````diff
@@ -0,0 +1,30 @@
+```
+
+### Superfluous member 
+A member in the etcd cluster can be considered superfluous only if either of the following 2 cases is true
````
```suggestion
A member in the etcd cluster can be considered superfluous only if either of the following 2 cases is true:
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
````diff
@@ -0,0 +1,30 @@
+Suppose an ETCD cluster is scaled up from `3` to `5` and the newly added members are failing constantly due to insufficient resources and then if the ETCD cluster is scaled back down to `3`. The failing member pods may not have the chance to clean up their `member` entries from the ETCD cluster leading to superfluous members in the ETCD cluster that may have an adverse effect on the quorum of the ...
+```
+
+### Superfluous member 
+A member in the etcd cluster can be considered superfluous only if either of the following 2 cases is true
+1. A member does not have a corresponding pod AND a corresponding lease
+2. A member is an unstarted learner and is part of the etcd member list without a name
````
```suggestion
A member in the etcd cluster is considered `superfluous` if either of the following 2 cases is true:
1. A member does not have a corresponding pod AND a corresponding lease
2. A member is an unstarted learner and is part of the etcd member list without a name
```

Please bring this to the top, just after the document title.

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+2. A member is an unstarted learner and is part of the etcd member list without a name
+
+### Approach
+- Superfluous members garbage collection can be enabled by setting the `enable-etcd-member-gc` flag to `true`
```
```suggestion
- Garbage collection of superfluous members can be enabled by setting the `enable-etcd-member-gc` flag to `true`
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+
+### Approach
+- Superfluous members garbage collection can be enabled by setting the `enable-etcd-member-gc` flag to `true`
+- The process is only triggered by the leading `etcd-backup-restore` sidecar and not on followers
```
```suggestion
- The process is only triggered by the leading `etcd-backup-restore` sidecar and not by the followers
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+- Superfluous members garbage collection can be enabled by setting the `enable-etcd-member-gc` flag to `true`
+- The process is only triggered by the leading `etcd-backup-restore` sidecar and not on followers
+- The garbage collection mechanism runs periodically with a default time period between runs set to `60s`, but can be modified with the `k8s-member-gc-duration` flag
+- It loops through the etcd cluster member list and verifies if a member is considered superfluous by checking if either of the 2 cases mentioned above holds true
```
```suggestion
- It loops through the etcd cluster member list and verifies if a member is considered superfluous by checking if either of the 2 conditions mentioned above for a member to be considered `superfluous` holds true
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+## Superfluous member deletion in the ETCD cluster
```
```suggestion
# Garbage Collection of Superfluous Members in the Etcd Cluster
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+The purpose of this document is the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check
+
+
+#### Example
```
```suggestion
## Example
```

shreyas-s-rao commented on docs/usage/etcd_member_garbage_collection.md:
```diff
@@ -0,0 +1,30 @@
+1. A member does not have a corresponding pod AND a corresponding lease
+2. A member is an unstarted learner and is part of the etcd member list without a name
+
+### Approach
```
```suggestion
## Approach
```

shreyas-s-rao reviewed (changes requested):
@aaronfern thanks for the PR, and apologies in the delay in reviewing this. Please address my comment, thanks.

shreyas-s-rao commented:
/ping @aaronfern

aaronfern commented:
`etcd-member-garbage-collection` will soon be removed from the codebase as it is not required anymore
Hence closing this PR as the doc will no longer be required 
/close
